flash_attn/cute/flash_fwd.py

```diff
@@ -1674,89 +1774,277 @@ def mma(
-                    tPrP = smem_thr_copy_P.retile(tOrP_cur)
-                    cute.copy(smem_thr_copy_P, tPrP, tPsP)
-                    # Fence and barrier to make sure smem store is visible to WGMMA
-                    cute.arch.fence_proxy(
-                        cute.arch.ProxyKind.async_shared, space=cute.arch.SharedSpace.shared_cta
-                    )
-                    cute.arch.sync_warp()  # Only need syncwarp since each warp is using its own P values for MmaPV
-                # Need to initialize tOrO in the case of RescaleOBeforeGemm where we will scale tOrO even in the 1st iter
-                # acc_O.fill(0.0)
-            else:
-                self.warp_scheduler_barrier_sync()
-                kv_consumer_state = mma_one_n_block(
-                    kv_consumer_state,
-                    n_block=n_block_max - 1,
-                    mma_pv_fn=partial(mma_pv_fn, zero_init=True),
-                    is_first_n_block=True,
-                    mask_fn=partial(mask_fn, mask_seqlen=True),
-                )
-                O_should_accumulate = True
-            # if cute.arch.thread_idx()[0] == 128: cute.printf("m_block = {}, n_block_max = {}, n_block_min = {}", m_block, n_block_max, n_block_min)
-            n_block_max -= 1
-            # Next couple of iterations with causal masking
-            if const_expr(self.is_causal or self.is_local):
-                n_block_min_causal_local_mask = block_info.get_n_block_min_causal_local_mask(
-                    seqlen, m_block, n_block_min
-                )
-                # if cute.arch.thread_idx()[0] == 128: cute.printf("n_block_min_causal_local_mask = {}", n_block_min_causal_local_mask)
-                for n_tile in cutlass.range(n_block_max - n_block_min_causal_local_mask, unroll=1):
+            
+            
+            # ==========================================
+            # MAINLOOP 
+            # ==========================================
+            if const_expr(mask_block_cnt is not None and full_block_cnt is not None):
+                # ==========================================
+                # Flex Attention blocksparsity
+                # ==========================================
+                curr_mask_block_cnt = mask_block_cnt[batch_idx, head_idx, m_block]
+                curr_full_block_cnt = full_block_cnt[batch_idx, head_idx, m_block]
+                if curr_mask_block_cnt > 0:
+                    # ==========================================
+                    # Masked blocks processing
+                    # ==========================================
+                    curr_mask_block_idx = mask_block_idx[batch_idx, head_idx, m_block, None]
+                    mask_n_block = 0
+                    if const_expr(self.intra_wg_overlap):
+                        mask_n_block = curr_mask_block_idx[curr_mask_block_cnt - 1]
+
+                        kv_consumer_state = self.first_half_block_overlap(
```

Comment:
Can we use functools.partial to make `first/last_half_block_overlap` shorter to call?